### PR TITLE
Trn 487a

### PR DESCRIPTION
**Pull request by pchristopher1275**
The first part of https://jira2.iparadigms.com/browse/TRN-487: which reads

As a walker operator I want domain priorities to work relatively linear to each other. If I have domains of priority 1, 2, and 3, the 3 should run (on average) three times as many as the 1, and the 2 should run (on average) twice as much as the 1. I want priorities to be integers, minimum 1, maximum whatever we can store in an unsigned integer.
Implementation notes to complete this work:
add a table domain_counters, keyed on domain and having a counter_type column next_crawl (this is a separate table because counters cannot exist next to non-counter non-primary-keys)
add a table walker_globals with columns key text and val int to keep global shared values
in the dispatcher kick off a separate goroutine that periodically (once per dispatch_interval) reads all priorities in domain_info, finds the max, and updates walker_globals with a row: key='max_priority', val=<the number> (maybe add a hard minimum on how frequently we'll get this value, say, 1 minute, in case people set dispatch_interval to 0 or very low)
add a private datastore method to get the current maxPriority from walker_globals, which should cache it for 1 minute and re-fetch it on demand. If this value does not exist in walker_globals, use default_domain_priority
in the fetcher's datastore, select all unclaimed and dispatched domains regardless of priority. For each one, increment the domain_counters.next_crawl by that domain's priority, then select it.
If the number is >= maxPriority, try to claim this domain. If we could not claim it, move on. If we claimed it, then decrement it's domain_counters.next_crawl by maxPriority
If it's next_crawl is not >= maxPriority move on, do not claim it
Note in the code that with multiple fetchmanagers going they could increment at the same time and read outdated next_crawl values (i.e. there are concurrency issues here), but that nothing bad will happen, at worst a domain will not be crawled on perfectly even intervals
in the query for unclaimed and dispatched hosts above, use paging so that a given fetcher iterates the whole domain_info table evenly. Keep the last domain this datastore incremented next_crawl, and for successive selects use token(dom) > token(<last domain>)
add to the FAQ explaining what priorities are for (high-level, for an operator)
update the console to allow setting priorities within the appropriate range
add a config key default_domain_priority and every time a domain is added set it to that. Default this key to 100
I believe that captures everything. Feel free to split this into multiple pull requests or tasks if necessary.

===

This patch only implements the part of the ticket that fixes the MaxPriority to 10 (thus the allowed priorities is 1, 2, ..., 10.

### CODE REVIEW

File: cassandra/datastore.go [hunk 1]
```diff
@@ -119,17 +116,68 @@ func (ds *Datastore) ClaimNewHost() string {
 	return domain
 }
 
+// domainPriorityTry will return true if the domain, dom, is eligible to be claimed.
+// The second argument, domPriority, is the domain priority of dom. Note this method updates
+// the domain_counters table.
+func (ds *Datastore) domainPriorityTry(dom string, domPriority int) bool {
```
Comment (dankinder) [inline cassandra/datastore.go]: For the sake of developers, could we add the note here about multiple crawlers going: "Note in the code that with multiple fetchmanagers going they could increment at the same time and read outdated next_crawl values (i.e. there are concurrency issues here), but that nothing bad will happen, at worst a domain will not be crawled on perfectly even intervals".

File: interfaces.go [hunk 2]
```diff
@@ -25,8 +25,10 @@ type Datastore interface {
 	// ClaimNewHost returns a hostname that is now claimed for this crawler to
 	// crawl. A segment of links for this host is assumed to be available.
 	// Returns the domain of the segment it claimed, or "" if there are none
-	// available.
-	ClaimNewHost() string
+	// available. seedDomain may be nil, or point to a string. If seedDomain
+	// is non-nil, the string pointed to will be read and written too, and a
+	// pointer to the same string should be passed to each call to ClaimNewHost.
+	ClaimNewHost(seedDomain *string) string
```
Comment (dankinder) [inline interfaces.go]: Is there a reason for seedDomain to be an argument to ClaimNewHost instead of an internally-managed variable in the Datastore? I was thinking it would be the latter.

File: cassandra/datastore.go (hunk 1 above)
Comment (pchristopher1275) [inline cassandra/datastore.go]: Yea I should have added a comment about this before y'all looked at it: I had focused on this bullet item in the original ticket. The reason I omitted the comment, is that I don't think we have a concurrency issue here (at least I wouldn't use that language). Consider the following setup, we have a domain with priority 8, with the max_priority of the universe at 10, and we have two different fetchers. Now consider these two situations

situation #1:
==========
* fetcher #1 increments next_crawl for our domain
* fetcher #2 increments next_crawl for our domain
* at this point next_crawl is 16 (> max_priority) so at-least one of the fetchers will get the domain
* [Important] one-and-only-one of these fetchers will be awarded the domain to crawl. This is enforced by tryClaimHosts which uses a compare-and-set operation to serialize access to the domain.
* either fetcher #1     OR    fetcher #2 will decrement next_crawl
* next_crawl is left at 8

Compare this list of events with:

situation #2:
==========
* fetcher #1 increments next_crawl for our domain, but next_crawl = 8 < max_priority so nothing happens
* time passes
* fetcher #2 increments next_crawl for our domain. So next_crawl = 16 > max_priority, and fetcher #2 will de facto win the CAS pull.
* fetcher #2 decrements next_crawl and wins the domain
* next_crawl is left at 8


The first situation is really similar to the second situation: but is the limit that fetcher #1 and fetcher #2 increment next_crawl at the same time. But because of the [important] step above: even if both fetcher #1 AND #2 see next_crawl as 16 (i.e. increment at the same time, and read the same twice incremented next_crawl), only ONE of them will be awarded the domain. And at the end of both fetchers trying to get the domain, the next_crawl will be the same (namely 8) in *either* situation. 

What do you think? Am I missing something you're concerned about?

Comment (dankinder) [inline cassandra/datastore.go]: Nope I think you've got it, I guess saying there are "concurrency issues" was unclear of me. Exactly as you described, I don't think we have any bad behavior possible here. The comment was to indicate this; i.e. I imagine a developer looking at this and thinking "hey, some other datastore could increment this at the same time, won't that be a problem?" and it would be good to have a brief little 2-3 line explanation so they know we've considered that and it's not a problem (because of the CAS claim and such, as you say).

File: interfaces.go (hunk 2 above)
Comment (pchristopher1275) [inline interfaces.go]: Oh yea, I forgot I was supposed to change that. Misunderstanding early gave rise to this design, and then I forgot to go back and fix. I'll patch it ...

File: cassandra/datastore.go (hunk 1 above)
Comment (pchristopher1275) [inline cassandra/datastore.go]: Gotcha. I agree that this code can be tricky, and that a comment is useful. I added a comment above domainPriorityTry that reads:

NOTE: next_crawl uses cassandra counters which can increment/decrement in a concurrent-consistent manner. Plus, the compare-and-set operation in tryClaimHosts guarantees that only one thread can claim a domain, even if several workers, on several different machines, are simultaneously trying to claim the domain. 

I hope that this will help the reader understand the code, and put their minds at ease. What do y'all think?

File: cmd/cmd_test.go [hunk 3]
```diff
@@ -117,7 +117,7 @@ func TestFetchCommand(t *testing.T) {
 	Handler(handler)
 
 	datastore := &helpers.MockDatastore{}
-	datastore.On("ClaimNewHost").Return("")
+	datastore.On("ClaimNewHost", mock.Anything).Return("")
```
Comment (dankinder) [inline cmd/cmd_test.go]: Think you'll need to undo these changes for mocking ClaimNewHost

**Comment by dankinder** [comment]
Yeah looks good. Outside of final change for the mock calls, :facepunch:
